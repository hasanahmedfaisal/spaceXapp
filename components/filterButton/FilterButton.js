import React from 'react'
import PropTypes from 'prop-types'
import isEmpty from 'lodash/isEmpty'
import * as QueryStringHelper from '../../helpers/QueryStringHelper'

const propTypes = {
  name: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  isSelected: PropTypes.bool,
  filterType: PropTypes.string,
  selectedFilters: PropTypes.objectOf(PropTypes.string)
}

const defaultProps = {
  isSelected: false
}

export default class FilterButton extends React.Component {
  constructor () {
    super()
    this.handleClick = this.handleClick.bind(this)
  }

  handleClick () {
    const selectedFilter = { ...this.props.selectedFilters }
    if (isEmpty(this.props.selectedFilters)) {
      selectedFilter[this.props.filterType] = this.props.name
    } else {
      if (!!(this.props.filterType in this.props.selectedFilters) &&
                this.props.selectedFilters[this.props.filterType] === this.props.name) {
        delete selectedFilter[this.props.filterType]
      } else {
        selectedFilter[this.props.filterType] = this.props.name
      }
    }
    this.props.onHandleClick(selectedFilter)
    if (isEmpty(selectedFilter)) { window.history.pushState({}, '', '/') } else { window.history.pushState({}, '', `filters?${QueryStringHelper.jsonToQueryString(selectedFilter)}`) }
  }

  render () {
    const buttonClassName = this.props.isSelected ? 'button isSelected' : 'button'
    return (
      <div className='buttonDiv'>
        <button className={buttonClassName} onClick={this.handleClick}>{this.props.name}</button>
      </div>
    )
  }
}

FilterButton.propTypes = propTypes
FilterButton.defaultProps = defaultProps
