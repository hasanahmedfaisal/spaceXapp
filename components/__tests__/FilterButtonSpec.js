import FilterButton from '../filterButton/FilterButton'
import renderer from 'react-test-renderer'

describe('Filter Button',() => {
    const { location } = window;

  beforeAll(() => {
    delete window.location;
    window.location = { assign: jest.fn() };
  });

  afterEach(() => {
      jest.clearAllMocks()
  })

  afterAll(() => {
    window.location = location;
  });

    it('should render correctly',() => {
        const filterButtonProps = {
            name: '2006',
            filterType: 'launch_year',
            isSelected: true,
            selectedFilters: {
                "launch_year": '2010'
            }
        }
        const tree = renderer.create(<FilterButton {...filterButtonProps}/>).toJSON()
        expect(tree).toMatchSnapshot()
    }) 

   it('should handle onClick with the filter type when selected filters is empty',() => {
        const filterButtonProps = {
            name: '2006',
            filterType: 'launch_year',
            isSelected: true,
            selectedFilters: {}
        }
        const root = renderer.create(<FilterButton {...filterButtonProps}/>).root
        root.findByType("button").props.onClick()
        expect(window.location.assign).toHaveBeenCalledWith('filters?launch_year=2006')
    }) 

    it('should handle onClick with no filter when selected filter and on click filter is equal',() => {
        const filterButtonProps = {
            name: '2006',
            filterType: 'launch_year',
            isSelected: true,
            selectedFilters: {
                launch_year: '2006'
            }
        }
        const root = renderer.create(<FilterButton {...filterButtonProps}/>).root
        root.findByType("button").props.onClick()
        expect(window.location.assign).toHaveBeenCalledWith('/')
    }) 

    it('should handle onClick with all filters when selected filter and on click filter is not equal',() => {
        const filterButtonProps = {
            name: '2006',
            filterType: 'launch_year',
            isSelected: true,
            selectedFilters: {
                launch_success: 'True',
                land_success: 'True'
            }
        }
        const root = renderer.create(<FilterButton {...filterButtonProps}/>).root
        root.findByType("button").props.onClick()
        expect(window.location.assign).toHaveBeenCalledWith('filters?launch_success=True&land_success=True&launch_year=2006')
    }) 
})