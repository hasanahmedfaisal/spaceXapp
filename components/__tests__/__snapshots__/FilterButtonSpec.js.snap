// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Button should render correctly 1`] = `
<div
  className="buttonDiv"
>
  <button
    className="button isSelected"
    onClick={[Function]}
  >
    2006
  </button>
</div>
`;
